### PR TITLE
[STORM-2440] Prevent KafkaSpout from hanging under certain Kafka failure scenarios

### PR DESCRIPTION
**Pull request by nicom**
Issue: https://issues.apache.org/jira/browse/STORM-2440

### CODE REVIEW

**Comment by srdo** [comment]
+1, nice catch.

Here's the master version, for reference https://github.com/apache/storm/pull/1932.

File: storm-core/src/clj/org/apache/storm/daemon/executor.clj [hunk 1]
```diff
@@ -267,9 +267,11 @@
                                ((:report-error <>) error)
                                (catch Exception e
                                  (log-message "Error while reporting error to cluster, proceeding with shutdown")))
-                             (if (or
-                                    (exception-cause? InterruptedException error)
-                                    (exception-cause? java.io.InterruptedIOException error))
+                             (if (and
```
Comment (HeartSaVioR) [inline storm-core/src/clj/org/apache/storm/daemon/executor.clj]: https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/executor/error/ReportErrorAndDie.java#L41-L43

According to current implementation for master branch, it should be

```
(if (or
  (exception-cause? InterruptedException error)
  (and
    (exception-cause? java.io.InterruptedIOException error)
    (not (exception-cause? java.net.SocketTimeoutException error))))))
```

**Comment by HeartSaVioR** [comment]
Please also change title to STORM-2440 instead of DEV-2440.

**Comment by nicom** [comment]
I just found out that this is a duplicate of #1932

**Comment by HeartSaVioR** [comment]
@nicom 
#1932 just addressed it for master branch so this PR is still valid. 
Could you double check my review comment and reopen / address?

File: storm-core/src/clj/org/apache/storm/daemon/executor.clj (hunk 1 above)
Comment (nicom) [inline storm-core/src/clj/org/apache/storm/daemon/executor.clj]: Fixed

**Comment by HeartSaVioR** [comment]
@nicom +1 Thanks for the quick fix.

**Comment by HeartSaVioR** [comment]
This PR is merged into asf git but asf git seems have some issue which might be related to mirror trigger. I'll try sync up mirror tomorrow. This will be closed automatically after sync up.

**Pull request merged**